src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp

```diff
@@ -0,0 +1,271 @@
+{
+  Callback _callback;
+
+  template <class... Args>
+  KOKKOS_FUNCTION void operator()(Args &&...args) const
+  {
+    _callback((Args &&) args...);
+  }
+};
+
+template <class Callback>
+auto declare_callback_constrained(Callback const &callback)
+{
+  return ConstrainedDistributedNearestCallback<Callback>{callback};
+}
+
+} // namespace Experimental
+
+namespace Details
+{
+
+struct DefaultCallbackWithRank
+{
+  int _rank;
+
+  template <typename Predicate, typename Value, typename OutputFunctor>
+  KOKKOS_FUNCTION void operator()(Predicate const &, Value const &value,
+                                  OutputFunctor const &out) const
+  {
+    out({value, _rank});
+  }
+};
+
+template <class Callback>
+struct is_constrained_callback : std::false_type
+{};
+template <class Callback>
+struct is_constrained_callback<
+    Experimental::ConstrainedDistributedNearestCallback<Callback>>
+    : std::true_type
+{};
+template <>
+struct is_constrained_callback<DefaultCallback> : std::true_type
+{};
+template <>
+struct is_constrained_callback<DefaultCallbackWithRank> : std::true_type
+{};
+
+template <class Callback>
+inline constexpr bool is_constrained_callback_v =
+    is_constrained_callback<Callback>::value;
+
+template <class Predicates, class Distances>
+struct WithinDistanceFromPredicates
+{
+  Predicates predicates;
+  Distances distances;
+};
+} // namespace Details
+
+template <class Predicates, class Distances>
+struct AccessTraits<
+    Details::WithinDistanceFromPredicates<Predicates, Distances>, PredicatesTag>
+{
+  using Predicate = typename Predicates::value_type;
+  using Geometry =
+      std::decay_t<decltype(getGeometry(std::declval<Predicate const &>()))>;
+  using Self = Details::WithinDistanceFromPredicates<Predicates, Distances>;
+
+  using memory_space = typename Predicates::memory_space;
+  using size_type = decltype(std::declval<Predicates const &>().size());
+
+  static KOKKOS_FUNCTION size_type size(Self const &x)
+  {
+    return x.predicates.size();
+  }
+  template <class Dummy = Geometry,
+            std::enable_if_t<std::is_same_v<Dummy, Geometry> &&
+                             std::is_same_v<Dummy, Point>> * = nullptr>
+  static KOKKOS_FUNCTION auto get(Self const &x, size_type i)
+  {
+    auto const point = getGeometry(x.predicates(i));
+    auto const distance = x.distances(i);
+    return intersects(Sphere{point, distance});
+  }
+  template <class Dummy = Geometry,
+            std::enable_if_t<std::is_same_v<Dummy, Geometry> &&
+                             std::is_same_v<Dummy, Box>> * = nullptr>
+  static KOKKOS_FUNCTION auto get(Self const &x, size_type i)
+  {
+    auto box = getGeometry(x.predicates(i));
+    auto &min_corner = box.minCorner();
+    auto &max_corner = box.maxCorner();
+    auto const distance = x.distances(i);
+    for (int d = 0; d < 3; ++d)
+    {
+      min_corner[d] -= distance;
+      max_corner[d] += distance;
+    }
+    return intersects(box);
```

Comment:
This enlarging the box can yield false positive right?  Is that potentially an issue?
We might want to add a comment.

Reply:
The behavior is unchanged since `master`. It can do false positive for the upper tree, resulting in more work to be done. From a user perspective, there is no false positives.